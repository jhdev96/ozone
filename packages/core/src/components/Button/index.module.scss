@import '../../scss/root';
@import '../../scss/mixins/button/button.scss';
@import '../../scss/mixins/button/variables.scss';
                                  

// For styling `flat` buttons
$flatMap: (
  primary: (
    color: #fff,
    bg-color: var(--blue-3)
  ),
  secondary: (
    color: #fff,
    bg-color: var(--red-3)
  ),
  warning: (
    color: var(--black-core),
    bg-color: var(--yellow-3)
  )
);

/* Major button classes - primary, secondary, warning, basic */                                   
.primary {
  @include button;

  color: #fff;
  background-color: var(--blue-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;

    color: var(--blue-core);
    border: 1px solid var(--blue-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--red-core) auto 1px !important;
  }
}

.secondary {
  @include button;

  color: #fff;
  background-image: $secondary-gradient;
  background-repeat: repeat-x;
  border-color: $button-bc1 $button-bc2 $button-bc3 !important;
  box-shadow: inset 0 1px 0 $primary-shade1, 
                    0 1px 2px $primary-shade2;
  background-color: var(--red-core);
  border-radius: 4px;

  &--outline {
    @extend .flat;
    color: var(--red-core);
    border: 1px solid var(--red-core) !important;
    background-color: transparent;
  }

  &:focus {
    outline: var(--blue-core) auto 1px !important;
  }
}

.warning {
  @include button;

  color: var(--black-core);
  background-color: var(--yellow-core);
  border-radius: 4px;

  &:focus {
    outline: var(--black-core) auto 1px !important;
  }
}

.basic {
  @include button;

  border-radius: 3px;

  &:focus {
    outline: var(--gray-4) auto 1px !important;
  }
}

/* The following styles serve as modifiers 
 * for the major button classes 
*/
.flat {
  background-image: none !important;
  background-repeat: no-repeat;
  border: 0;
  border-radius: 0 !important;
  box-shadow: none !important;
  text-shadow: none !important;

  @each $class, $rule in $flatMap {
    &.#{$class}:not(.disabled):not(.loading) {
      &:hover {
        color: map-get($rule, 'color');
        background-color: map-get($rule, 'bg-color');
      }
    }
  }
}

.fill { width: 100%; }

.loading {
  padding: 8px 20px 5px 20px;
  max-height: 35px;
  opacity: 0.90;
  cursor: default;

  &:focus {outline: none !important;}
}

.disabled {
  opacity: 0.70;
  cursor: default;
}

.small {
  min-width: fit-content;
  padding: 8px !important;
}

.medium { min-width: 40px; }

.large {
  min-width: 70px;
  max-height: 70px !important;
  padding: 15px 20px !important;
  font-size: 1.2em !important;
}

.light {
  &.basic {
    color: var(--black-core);
    background-color: #fff;
    border: 1px solid var(--gray-3);
  }
}

.dark {
  &.basic {
    color: var(--gray-1);
    border: 1px solid var(--black-6);
    background-color: var(--black-4) !important;
  }
}