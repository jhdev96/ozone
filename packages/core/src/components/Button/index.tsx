import React from 'react';
import classNames from '../../utils/classNames';
import { 
  ButtonVariation,
  loaderTheme
} from './utils';
import LoadingDots from '../LoadingDots';
import styles from './index.module.scss';


export interface ButtonProps {
  /** 
   * The main classes of buttons. 
   */
  variation?: ButtonVariation;
  /** 
   * Sets the HTML `type` attribute for the
   * button element.
   */ 
  type?: React.ButtonHTMLAttributes<HTMLButtonElement>["type"]
  /** 
   * Control the size of the button. Default is medium. 
   */
  size?: "small" | "medium" | "large";
  /** 
   * Give the button a flat texture, appropriate for forms. 
   */
  flat?: boolean;
  /** 
   * When true the button will no background; 
   * It will only have a border. 
   */
  outline?: boolean;
  /** 
   * When true, the button's width will be 100% of 
   * its container. 
   */ 
  fill?: boolean;
  /** 
   * Color the button based on the current theme. 
   */
  theme?: "light" | "dark";
  /** 
   * Make the button disabled. 
   */
  isDisabled?: boolean;
  /** 
   * Indicates whether the button is loading some content. 
   */
  isLoading?: boolean;
  /** 
   * Sets the `className` attribute on the button element. 
   */ 
  className?: string;
  /** 
   * The function that handles click events. 
   */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers over the button.
   */
  onMouseEnter?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Called when the user's mouse hovers away from the
   * button.
   */
  onMouseLeave?: React.MouseEventHandler<HTMLButtonElement>;
  /** 
   * A selector used for testing this component. 
   */
  dataTestId?: string;
  /** 
   * Content displayed inside the button. 
   */
  children?: React.ReactNode | string;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button({
    variation     = "primary",
    type          = "button",
    size          = "medium", 
    flat          = false,
    outline       = false,
    fill          = false, 
    theme         = "light",
    isDisabled    = false,
    isLoading     = false,
    className,
    onClick       = () => {},
    onMouseEnter  = () => {},
    onMouseLeave  = () => {},
    dataTestId,
    children}: ButtonProps,
    ref: React.Ref<HTMLButtonElement>): JSX.Element 
  {
    const childMarkup = isLoading 
      ? (
          <>
            <LoadingDots 
              dotSize={7} 
              position="center" 
              theme={loaderTheme(variation, theme, outline)} 
            />
            <span style={{opacity: '0'}}>{children}</span>
          </>
        )
      : children;

    return (
      <button
        type={type}
        className={classNames(
          className,
          styles[variation],
          styles[size],
          styles[theme],
          flat        && styles.flat,
          outline     && styles[`${variation}--outline`],
          fill        && styles.fill,
          isLoading   && styles.loading,
          isDisabled  && styles.disabled
        )}
        onClick={isDisabled ? () => {} : onClick}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        disabled={isDisabled}
        aria-busy={isLoading}
        data-testid={dataTestId}
        ref={ref}
      >
        {childMarkup}
      </button>
    );
  }
);

Button.displayName = 'Button';

export default Button;