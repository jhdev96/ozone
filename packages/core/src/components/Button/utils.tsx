import { LoadingDotsProps } from '../LoadingDots';


export type ButtonVariation = "primary" | "secondary" | "warning" | "basic";

export const loaderTheme = (
  btnType: ButtonVariation, 
  theme: "light" | "dark", 
  outline: boolean
): LoadingDotsProps["theme"] => {
  /** 
   * Sets the theme for the LoadingDots component
   * which is rendered as a child for a button
   * that is loading. 
   * 
   * @param btnType // the type of button wrapping the loader
   * @param theme // the button's theme
   * @param outline // whether the button has just an outline
   * 
   * This function will return
   * an appropriate theme for the LoadingDots
   * component, so that it contrasts with the
   * button.
   */
  switch(btnType) {
    case 'basic':
      if (theme === 'light') return 'dark';
      return 'light';
    case 'primary':
      if (outline) return "blue";
      return theme;
    case 'secondary':
      if (outline) return "red";
      return theme;
    case 'warning':
      return "dark";
    default:
      return theme;
  }
}