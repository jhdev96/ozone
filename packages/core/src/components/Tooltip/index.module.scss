@import '../../scss/root';
@import '../../scss/mixins/popover/popover.scss';


.tooltipWrapper {
  @include wrapper;
}

.tooltip {
  @include popover;
  
  color: #fff;
  background: var(--blue-7);
  font-size: 12px;
  line-height: 1;
  opacity: 1;
  cursor: default;
}

/* Position */
.top {
  top: calc(var(--tooltipMargin) * -3);
}

.bottom {
  bottom: calc(var(--tooltipMargin) * -3);
}

.left {
  @include horizontalPosition(left);
}

.right {
  @include horizontalPosition(right);
}

/* State */
.fadeOut {
  opacity: 0;
  transition: opacity 0.3s;
}