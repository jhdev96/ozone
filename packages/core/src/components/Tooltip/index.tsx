import React, { useState, useEffect } from 'react';
import { PopoverProps } from '../Popover';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';


interface TooltipProps extends PopoverProps {
  /** 
   * The amount of time in milliseconds to wait
   * before showing or hiding the tooltip. 
   */ 
  delay?: number;
  /**
   * If true, the tooltip will fade in and out
   * when its visibility changes.
   */
  animate?: boolean;
}

export default function Tooltip({
  position  = "top",
  delay     = 500,
  content,
  animate   = true,
  isVisible = false,
  onShow    = () => {},
  onHide    = () => {},
  dataTestId,
  children
}: TooltipProps): JSX.Element 
{
  const [visible, setVisible] = useState(isVisible);
  const [fadeOut, setFadeOut] = useState(false);

  let delayFn: ReturnType<typeof setTimeout>;

  const showTooltip = () => {
    delayFn = setTimeout(() => {
      setFadeOut(false);
      setVisible(true);
      onShow();
    }, delay);
  }

  const hideTooltip = () => {
    delayFn = setTimeout(() => {
      setFadeOut(true)
      onHide();
    }, delay);
  }

  useEffect(() => {
    if (fadeOut && visible) {
      delayFn = setTimeout(() => {
        setVisible(false);
      }, 300);
    }
    return () => clearTimeout(delayFn);
  }, [fadeOut])

  return (
    <div 
      className={styles.tooltipWrapper}
      onMouseEnter={showTooltip}
      onMouseLeave={hideTooltip}
      data-testid={dataTestId}
    >
      {children}
      {visible && (
        <div 
          className={classNames(
            styles.tooltip, 
            styles[position],
            animate && fadeOut && styles.fadeOut
          )}
        >
          {content}
        </div>
      )}
    </div>
  )
}

Tooltip.displayName = 'Tooltip';