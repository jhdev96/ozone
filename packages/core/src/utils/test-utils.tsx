import React from "react";


interface TestProps {
  dataTestId?: string;
}

export function DummyTestComponent({
  dataTestId = "test"
}: TestProps): JSX.Element 
{
  return (
    <span data-testid={dataTestId}>
      Hello
    </span>
  )
}

export function mockRefCurrent(currentProperties?: object) 
{
  const ref = {current: currentProperties};
  
  jest.spyOn(React, 'useRef').mockImplementation(() => {
    return ref;
  });
}