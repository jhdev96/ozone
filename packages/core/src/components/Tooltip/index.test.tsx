import React from "react";
import { 
  render, 
  screen, 
  fireEvent
} from "@testing-library/react";  
import { act } from "react-dom/test-utils";
import Tooltip from ".";
import Button from '../Button';


beforeEach(() => {
  /** 
   * <Tooltip /> uses 'setTimeout' in 
   * its useEffect hook. This ensures
   * that 'setTimeout' gets mocked by
   * jest.
   */
  jest.useFakeTimers();
});

afterEach(() => {
  jest.useRealTimers();
});

describe('<Tooltip />', () => {
  it('renders the target element', () => {
    render(
      <Tooltip>
        <Button dataTestId="button">
          Hover over me
        </Button>
      </Tooltip>
    );

    expect(screen.getByTestId('button'))
      .toBeInTheDocument;
  });

  it('renders an invisible tooltip by default', () => {
    render(
      <Tooltip>
        <span>text</span>
      </Tooltip>
    );

    const tooltip = document.querySelector('.tooltip');

    expect(tooltip).not.toBeInTheDocument();
  });

  it('renders a visible tooltip when `isVisible` is true', () => {
    render(
      <Tooltip isVisible>
        <span>text</span>
      </Tooltip>
    );

    const tooltip = document.querySelector('.tooltip');

    expect(tooltip).toBeInTheDocument();
  });

  it('renders a tooltip with some content', () => {
    render(
      <Tooltip content="tooltip" isVisible>
        <span>text</span>
      </Tooltip>
    );

    const tooltip = document.querySelector('.tooltip');

    expect(tooltip).toHaveTextContent('tooltip')
  });

  it('renders the tooltip when the mouse hovers over the target', () => {
    act(() => {
      render(
        <Tooltip 
          dataTestId="test" 
          content="tooltip" 
        >
          <span>text</span>
        </Tooltip>
      );

      fireEvent.mouseEnter(screen.getByTestId('test'));
      jest.runAllTimers();
    });

    const tooltip = document.querySelector('.tooltip');
  
    expect(tooltip).toBeInTheDocument();
  });

  it('calls `onShow` when the tooltip is visible', () => {
    const onShow = jest.fn();

    act(() => {
      render(
        <Tooltip 
          dataTestId="test" 
          content="tooltip" 
          onShow={onShow}
        >
          <span>text</span>
        </Tooltip>
      );

      fireEvent.mouseEnter(screen.getByTestId('test'));
      jest.runAllTimers();
    });

    expect(onShow).toBeCalled();
  });

  it('calls `onHide` when the tooltip loses visibility', () => {
    const onHide = jest.fn();

    act(() => {
      render(
        <Tooltip 
          dataTestId="test" 
          content="tooltip" 
          onHide={onHide}
          isVisible
        >
          <span>text</span>
        </Tooltip>
      );

      fireEvent.mouseLeave(screen.getByTestId('test'));
      jest.runAllTimers();
    });

    expect(onHide).toBeCalled();
  });

  it('sets the correct position', () => {
    const {container} = render(
      <>
        <Tooltip position="top" isVisible>text</Tooltip>
        <Tooltip position="bottom" isVisible>text</Tooltip>
        <Tooltip position="left" isVisible>text</Tooltip>
        <Tooltip position="right" isVisible>text</Tooltip>
      </>
    );

    expect(container).toMatchSnapshot();
  });
});