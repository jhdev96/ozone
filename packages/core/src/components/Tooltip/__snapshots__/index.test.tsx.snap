// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> sets the correct position 1`] = `
<div>
  <div
    class="tooltipWrapper"
  >
    text
    <div
      class="tooltip top"
    />
  </div>
  <div
    class="tooltipWrapper"
  >
    text
    <div
      class="tooltip bottom"
    />
  </div>
  <div
    class="tooltipWrapper"
  >
    text
    <div
      class="tooltip left"
    />
  </div>
  <div
    class="tooltipWrapper"
  >
    text
    <div
      class="tooltip right"
    />
  </div>
</div>
`;
